sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -981,6 +982,10 @@ object EliminateSorts extends Rule[LogicalPlan] {
       j.copy(left = recursiveRemoveSort(originLeft), right = recursiveRemoveSort(originRight))
     case g @ Aggregate(_, aggs, originChild) if isOrderIrrelevantAggs(aggs) =>
       g.copy(child = recursiveRemoveSort(originChild))
+    case r: RepartitionByExpression =>
+      r.copy(child = recursiveRemoveSort(r.child))
```

Comment:
Should we also limit to `partitionExpressions` are all deterministic case?

Reply:
I don't think that is necessary. My understanding is that the requirement for non-deterministic expressions is that they are run only once when finally executed, not that the plan cannot be optimized around them.

Reply:
Non-deterministic expressions could have status and row-order sensitive. In `EliminateSorts` we also only remove Sort for Project and Filter where projection list and condition are deterministic.

Reply:
You're right, that does seem to imply that ordering is preserved underneath non-deterministic expressions.

Reply:
Agree, let me update this.

Reply:
Updated.

Reply:
Could you update the PR description accordingly?

Reply:
Done.